### ISSUE TITLE
Parsing HTML and then pretty printing it

### ISSUE THREAD

**Issue by rlaferla** [cezheng/Fuzi#18]
I'm trying to parse some html text and then pretty print the entire document.  I couldn't tell what was the best way to traverse the hierarchy of nodes/elements and wasn't sure how to get the inner html content of a tag.  I'm posting here because I think this could improve the documentation for the API.

```
   let html = "**** put some html text here. ****"

            let doc = try HTMLDocument(string: html, encoding: NSUTF8StringEncoding)
            
            if let root = doc.root {
                let str = self.dumpElement(root)
                print(str)
            }

    func dumpElement(element:XMLElement) -> String {
        var str = ""
        
        str = "<\(element.tag!.uppercaseString)"
        for attr in element.attributes {
            str += " \(attr.0)=\(attr.1)"
        }
        str += ">"
        let nodes = element.childNodes(ofTypes: [.Text])
        for node in nodes {
            str += node.stringValue
        }
        
        for el in element.children {
            str += self.dumpElement(el)
        }
        str += "</\(element.tag!.uppercaseString)>"
        return str
    }
```

Is this correct?

**Comment by cezheng** [comment]
What do you mean exactly by `pretty print`?

The code you shared actually does the following things:
* Make the element's tag name uppercase
* Ignored all child node that are not text nodes and elements(which includes CDATA, comment, etc.)
* Dump all text nodes first, then elements after them, regardless of their order

So I don't really think it makes sense.

**Comment by cezheng** [comment]
I guess you only want to `pretty print` some xml for while debugging? try the `xmllint` command.

**Comment by hvtor** [comment]
How might you create the html String?  I'm using a separate class that inherits from NSObject to parse down a URL.

func httpGet(request: NSURLRequest!, callback: @escaping (String, String?) -> Void) {
        var session = URLSession.shared
        var task = session.dataTask(with: request as URLRequest){
            (data, response, error) -> Void in
            if error != nil {
                callback("error", error?.localizedDescription)
            } else {
                var result = String(data: data!, encoding:
                    String.Encoding(rawValue: String.Encoding.ascii.rawValue))!
                callback(result as String, nil)
            }
        }
        task.resume()
    }


in my ViewController I'm trying to : 

let html = data
do {
        let doc = try HTMLDocument(string: html, encoding: String.Encoding.utf8)
        } catch {   
        }

but I get an error for the 'html' variable. 

How do I set the html as a string from my initial URLrequest set up in my data service?

**Comment by cezheng** [comment]
@hvtor you don't need to create the string if you have NSData. It is stated in the README that you can create a document with either a String, an NSData(Data for Swift 3), or [CChar] instance. Actually having a NSData instance is simpler since you don't have to specify the Encoding.

```swift
let doc = try HTMLDocument(data: data)
```

**Comment by hvtor** [comment]
@cezheng Yes. Thank you. :) BTW, great documentation. Just a bit 😴 I guess.

**Comment by cezheng** [comment]
@hvtor haha, it's true. Any suggestions on improving it?

**Comment by hvtor** [comment]
No I meant I am 😴. It's great documentation. An example webpage would be great too! Showing how the elements can be mapped over. I'm not too familiar with 

I'm trying to parse an IMDb list and it's a series of anchor tags. It's not clear to me how you select for specific tags. 
<td class = "column_title">
<a>Walking Dead </a>
<a>Stranger Things</a>


*goes to read the docs again*


Parent. And then the child tags.
